File: custos-client-sdks/custos-java-clients/agent-management-client/src/main/java/org/apache/custos/agent/management/client/AgentManagementClient.java
package org.apache.custos.agent.management.client;

import io.grpc.ManagedChannel;
import io.grpc.netty.GrpcSslContexts;
import io.grpc.netty.NettyChannelBuilder;
import io.grpc.stub.MetadataUtils;
import org.apache.custos.agent.management.service.AgentManagementServiceGrpc;
import org.apache.custos.agent.management.service.AgentRegistrationResponse;
import org.apache.custos.agent.management.service.AgentSearchRequest;
import org.apache.custos.clients.core.ClientUtils;
import org.apache.custos.iam.service.*;

import java.io.IOException;
import java.util.Arrays;

/**
 * This class contains methods of Agent Management
 */
public class AgentManagementClient {

    private ManagedChannel managedChannel;

    private AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub;


    public AgentManagementClient(String serviceHost, int servicePort, String clientId,
                                 String clientSecret) throws IOException {

        managedChannel = NettyChannelBuilder.forAddress(serviceHost, servicePort)
                .sslContext(GrpcSslContexts
                        .forClient()
                        .trustManager(ClientUtils.getServerCertificate(serviceHost, clientId, clientSecret)) // public key
                        .build())
                .build();

        blockingStub = AgentManagementServiceGrpc.newBlockingStub(managedChannel);
        blockingStub = MetadataUtils.attachHeaders(blockingStub, ClientUtils.getAuthorizationHeader(clientId, clientSecret));
    }


    /**
     * Enable agents, this is should be the first operation to enable agent functions
     * @param adminToken
     * @return
     */
    public OperationStatus enableAgents(String adminToken) {

        AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub =
                MetadataUtils.attachHeaders(this.blockingStub, ClientUtils.getAuthorizationHeader(adminToken));

        AgentClientMetadata metadata = AgentClientMetadata
                .newBuilder()
                .build();

        return blockingStub.enableAgents(metadata);
    }


    /**
     * Sets access token life time
     * @param adminToken
     * @param accessTokenLifeTime
     * @return
     */
    public OperationStatus configureAgentClient(String adminToken, long accessTokenLifeTime) {

        AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub =
                MetadataUtils.attachHeaders(this.blockingStub, ClientUtils.getAuthorizationHeader(adminToken));

        AgentClientMetadata metadata = AgentClientMetadata
                .newBuilder()
                .setAccessTokenLifeTime(accessTokenLifeTime)
                .build();

        return blockingStub.configureAgentClient(metadata);
    }

    /**
     * This will register agent and activate agent
     * @param adminToken
     * @param id
     * @param realmRoles
     * @param agentAttributes
     * @return
     */
    public AgentRegistrationResponse registerAndEnableAgent(String adminToken, String id, String[] realmRoles, UserAttribute[] agentAttributes) {

        AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub =
                MetadataUtils.attachHeaders(this.blockingStub, ClientUtils.getAuthorizationHeader(adminToken));

        UserRepresentation userRepresentation = UserRepresentation
                .newBuilder()
                .setId(id)
                .addAllRealmRoles(Arrays.asList(realmRoles))
                .addAllAttributes(Arrays.asList(agentAttributes))
                .build();

        RegisterUserRequest userRequest = RegisterUserRequest
                .newBuilder()
                .setUser(userRepresentation)
                .build();

        return blockingStub.registerAndEnableAgent(userRequest);

    }


    /**
     * Get agent
     * @param adminToken
     * @param agentId
     * @return
     */
    public Agent getAgent(String adminToken, String agentId) {
        AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub =
                MetadataUtils.attachHeaders(this.blockingStub, ClientUtils.getAuthorizationHeader(adminToken));

        AgentSearchRequest request = AgentSearchRequest
                .newBuilder()
                .setId(agentId)
                .build();

        return blockingStub.getAgent(request);
    }

    /**
     * Delete agent
     * @param adminToken
     * @param agentId
     * @return
     */
    public OperationStatus deleteAgent(String adminToken, String agentId) {
        AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub =
                MetadataUtils.attachHeaders(this.blockingStub, ClientUtils.getAuthorizationHeader(adminToken));

        AgentSearchRequest request = AgentSearchRequest
                .newBuilder()
                .setId(agentId)
                .build();

        return blockingStub.deleteAgent(request);
    }


    /**
     * Disable agent
     * @param adminToken
     * @param agentId
     * @return
     */
    public OperationStatus disableAgent(String adminToken, String agentId) {
        AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub =
                MetadataUtils.attachHeaders(this.blockingStub, ClientUtils.getAuthorizationHeader(adminToken));

        AgentSearchRequest request = AgentSearchRequest
                .newBuilder()
                .setId(agentId)
                .build();

        return blockingStub.disableAgent(request);

    }


    /**
     * Enable agent
     * @param adminToken
     * @param agentId
     * @return
     */
    public OperationStatus enableAgent(String adminToken, String agentId) {
        AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub =
                MetadataUtils.attachHeaders(this.blockingStub, ClientUtils.getAuthorizationHeader(adminToken));

        AgentSearchRequest request = AgentSearchRequest
                .newBuilder()
                .setId(agentId)
                .build();

        return blockingStub.enableAgent(request);
    }


    /**
     * Add agent attributes
     * @param adminToken
     * @param agents
     * @param attributes
     * @return
     */
    public OperationStatus addAgentAttributes(String adminToken, String[] agents, UserAttribute[] attributes) {
        AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub =
                MetadataUtils.attachHeaders(this.blockingStub, ClientUtils.getAuthorizationHeader(adminToken));

        AddUserAttributesRequest addUserAttributesRequest = AddUserAttributesRequest
                .newBuilder()
                .addAllAttributes(Arrays.asList(attributes))
                .addAllAgents(Arrays.asList(agents))
                .build();
        return blockingStub.addAgentAttributes(addUserAttributesRequest);

    }


    /**
     * Delete agent attributes
     * @param adminToken
     * @param agents
     * @param attributes
     * @return
     */
    public OperationStatus deleteAgentAttributes(String adminToken, String[] agents, UserAttribute[] attributes) {
        AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub =
                MetadataUtils.attachHeaders(this.blockingStub, ClientUtils.getAuthorizationHeader(adminToken));

        DeleteUserAttributeRequest addUserAttributesRequest = DeleteUserAttributeRequest
                .newBuilder()
                .addAllAttributes(Arrays.asList(attributes))
                .addAllAgents(Arrays.asList(agents))
                .build();
        return blockingStub.deleteAgentAttributes(addUserAttributesRequest);
    }


    /**
     * Add roles to agents
     * @param adminToken
     * @param agents
     * @param roles
     * @return
     */
    public OperationStatus addRolesToAgents(String adminToken, String[] agents, String[] roles) {
        AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub =
                MetadataUtils.attachHeaders(this.blockingStub, ClientUtils.getAuthorizationHeader(adminToken));

        AddUserRolesRequest addUserRolesRequest = AddUserRolesRequest
                .newBuilder()
                .addAllAgents(Arrays.asList(agents))
                .addAllRoles(Arrays.asList(roles))
                .build();
        return blockingStub.addRolesToAgent(addUserRolesRequest);

    }


    /**
     * Delete roles from agents
     * @param adminToken
     * @param id
     * @param roles
     * @return
     */
    public OperationStatus deleteRolesFromAgents(String adminToken, String id, String[] roles) {
        AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub =
                MetadataUtils.attachHeaders(this.blockingStub, ClientUtils.getAuthorizationHeader(adminToken));

        DeleteUserRolesRequest deleteUserRolesRequest = DeleteUserRolesRequest
                .newBuilder()
                .addAllRoles(Arrays.asList(roles))
                .setId(id)
                .build();
        return blockingStub.deleteRolesFromAgent(deleteUserRolesRequest);

    }

    /**
     * Add protocol mapper, this will add attributes and roles to token
     * @param adminToken
     * @param name
     * @param attributeName
     * @param claimName
     * @param claimType
     * @param mapperType
     * @param addToIdToken
     * @param addToAccessToken
     * @param addToUserInfo
     * @param multiValued
     * @param aggregrateAttributeValues
     * @return
     */
    public OperationStatus addProtocolMapper(String adminToken, String name, String attributeName, String claimName, String claimType, String mapperType,
                                             boolean addToIdToken, boolean addToAccessToken, boolean addToUserInfo, boolean multiValued,
                                             boolean aggregrateAttributeValues) {
        AgentManagementServiceGrpc.AgentManagementServiceBlockingStub blockingStub =
                MetadataUtils.attachHeaders(this.blockingStub, ClientUtils.getAuthorizationHeader(adminToken));

        AddProtocolMapperRequest mapperRequest = AddProtocolMapperRequest
                .newBuilder()
                .setName(name)
                .setAttributeName(attributeName)
                .setClaimName(claimName)
                .setClaimType(ClaimJSONTypes.valueOf(claimType))
                .setMapperType(MapperTypes.valueOf(mapperType))
                .setAddToIdToken(addToIdToken)
                .setAddToAccessToken(addToAccessToken)
                .setAddToUserInfo(addToUserInfo)
                .setMultiValued(multiValued)
                .setAggregateAttributeValues(aggregrateAttributeValues)
                .build();
        return blockingStub.addProtocolMapper(mapperRequest);


    }


}

File: custos-client-sdks/custos-python-sdk/custos/clients/agent_management_client.py
#  Licensed to the Apache Software Foundation (ASF) under one or more
#  contributor license agreements.  See the NOTICE file distributed with
#  this work for additional information regarding copyright ownership.
#  The ASF licenses this file to You under the Apache License, Version 2.0
#  (the "License"); you may not use this file except in compliance with
#  the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#


import logging
import grpc

from custos.transport.settings import CustosServerClientSettings

from custos.server.integration.AgentManagementService_pb2_grpc import AgentManagementServiceStub
from custos.server.core.IamAdminService_pb2 import AgentClientMetadata, RegisterUserRequest, \
    UserRepresentation, UserAttribute, AddUserAttributesRequest, DeleteUserAttributeRequest, AddUserRolesRequest, \
    DeleteUserRolesRequest, AddProtocolMapperRequest, ClaimJSONTypes, MapperTypes
from custos.server.integration.AgentManagementService_pb2 import AgentSearchRequest
from custos.clients.utils.certificate_fetching_rest_client import CertificateFetchingRestClient

logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)


class AgentManagementClient(object):

    def __init__(self, custos_server_setting):
        self.custos_settings = custos_server_setting
        self.target = self.custos_settings.CUSTOS_SERVER_HOST + ":" + str(self.custos_settings.CUSTOS_SERVER_PORT)
        certManager = CertificateFetchingRestClient(custos_server_setting)
        certManager.load_certificate()
        with open(self.custos_settings.CUSTOS_CERT_PATH, 'rb') as f:
            trusted_certs = f.read()
        self.channel_credentials = grpc.ssl_channel_credentials(root_certificates=trusted_certs)
        self.channel = grpc.secure_channel(target=self.target, credentials=self.channel_credentials)
        self.agent_stub = AgentManagementServiceStub(self.channel)

    def enable_agents(self, token):
        """
        Enable agent registration for realm
        :param token:
        :return:
        """
        try:
            token = "Bearer " + token
            metadata = (('authorization', token),)

            return self.agent_stub.enableAgents(metadata=metadata)
        except Exception:
            logger.exception("Error occurred while enabling agents")
            raise

    def configure_agent_client(self, token, access_token_life_time):
        """
        Configure agent client
        :param token:
        :param access_token_life_time:
        :return:
        """
        try:
            token = "Bearer " + token
            metadata = (('authorization', token),)

            request = AgentClientMetadata(access_token_life_time=access_token_life_time)

            return self.agent_stub.enableAgents(request, metadata=metadata)
        except Exception:
            logger.exception("Error occurred while configuring agent client")
            raise

    def register_and_enable_agent(self, token, agent):
        """
        Register and enable agent
        :param agent:
        :param token:
        :return:
        """
        try:
            token = "Bearer " + token
            metadata = (('authorization', token),)

            attributeList = []
            for atr in agent['attributes']:
                attribute = UserAttribute(key=atr['key'], values=atr['values'])
                attributeList.append(attribute)
            id = agent['id']
            realm_roles = agent['realm_roles']
            user = UserRepresentation(id=id, realm_roles=realm_roles, attributes=attributeList)
            request = RegisterUserRequest(user=user)

            return self.agent_stub.registerAndEnableAgent(request=request, metadata=metadata)

        except Exception:
            logger.exception("Error occurred while enabling agents")
            raise

    def get_agent(self, token, id):
        """
        Get agent having id
        :param token:
        :param id:
        :return:
        """
        try:
            token = "Bearer " + token
            metadata = (('authorization', token),)

            request = AgentSearchRequest(id=id)
            return self.agent_stub.getAgent(request=request, metadata=metadata)
        except Exception:
            logger.exception("Error occurred while fetching agent")
            raise

    def delete_agent(self, token, id):
        """
        Delete agent having id
        :param token:
        :param id:
        :return:
        """
        try:
            token = "Bearer " + token
            metadata = (('authorization', token),)

            request = AgentSearchRequest(id=id)
            return self.agent_stub.deleteAgent(request=request, metadata=metadata)
        except Exception:
            logger.exception("Error occurred while deleting agent")
            raise

    def disable_agent(self, token, id):
        """
        Disable agent having id
        :param token:
        :return:
        """
        try:
            token = "Bearer " + token
            metadata = (('authorization', token),)

            request = AgentSearchRequest(id=id)
            return self.agent_stub.disableAgent(request=request, metadata=metadata)
        except Exception:
            logger.exception("Error occurred while disabling agent")
            raise

    def enable_agent(self, token, id):
        """
        Enable agent having id
        :param token:
        :param id:
        :return:
        """
        try:
            token = "Bearer " + token
            metadata = (('authorization', token),)

            request = AgentSearchRequest(id=id)
            return self.agent_stub.enableAgent(request=request, metadata=metadata)
        except Exception:
            logger.exception("Error occurred while enabling agent")
            raise

    def add_agent_attributes(self, token, agents, attributes):
        """
        Add attributes to agents
        :param token:
        :param agents:
        :param attributes:
        :return:
        """
        try:
            token = "Bearer " + token
            metadata = (('authorization', token),)
            attributeList = []
            for atr in attributes:
                attribute = UserAttribute(key=atr['key'], values=atr['values'])
                attributeList.append(attribute)

            request = AddUserAttributesRequest(attributes=attributeList, agents=agents)
            return self.agent_stub.addAgentAttributes(request=request, metadata=metadata)
        except Exception:
            logger.exception("Error occurred while adding agent attributes")
            raise

    def delete_agent_attributes(self, token, agents, attributes):
        """
        Delete agent attributes of agents
        :param token:
        :param agents:
        :param attributes:
        :return:
        """
        try:
            token = "Bearer " + token
            metadata = (('authorization', token),)
            attributeList = []
            for atr in attributes:
                attribute = UserAttribute(key=atr['key'], values=atr['values'])
                attributeList.append(attribute)

            request = DeleteUserAttributeRequest(attributes=attributeList, agents=agents)
            return self.agent_stub.deleteAgentAttributes(request=request, metadata=metadata)
        except Exception:
            logger.exception("Error occurred while deleting agent attributes")
            raise

    def add_roles_to_agents(self, token, agents, roles):
        """
        Add roles to agents
        :param token:
        :param agents:
        :param roles:
        :return:
        """
        try:
            token = "Bearer " + token
            metadata = (('authorization', token),)
            request = AddUserRolesRequest(agents=agents, roles=roles)
            return self.agent_stub.addRolesToAgent(request=request, metadata=metadata)
        except Exception:
            logger.exception("Error occurred whiling adding roles to agents")
            raise

    def delete_roles_from_agent(self, token, id, roles):
        """
        Delete roles from agent
        :param token:
        :param id:
       :param roles:
       :return:
       """

        try:
            token = "Bearer " + token
            metadata = (('authorization', token),)
            request = DeleteUserRolesRequest(id=id, roles=roles)
            return self.agent_stub.deleteRolesFromAgent(request=request, metadata=metadata)
        except Exception:
            logger.exception("Error occurred while enabling agents")
            raise

    def add_protocol_mapper(self, token, name, attribute_name, claim_name, claim_type, mapper_type,
                            add_to_id_token, add_to_access_token, add_to_user_info, multi_valued,
                            aggregate_attribute_values):
        """
        Add protocol mapper to agent client
        :param token:
        :param name:
        :param attribute_name:
        :param claim_name:
        :param claim_type:
        :param mapper_type:
        :param add_to_id_token:
        :param add_to_access_token:
        :param add_to_user_info:
        :param multi_valued:
        :param aggregate_attribute_values:
        :return:
        """
        try:
            token = "Bearer " + token
            metadata = (('authorization', token),)

            wrapped_json_type = ClaimJSONTypes.Value(claim_type)
            wrapped_mapper_type = MapperTypes.Value(mapper_type)

            request = AddProtocolMapperRequest(name=name,
                                               attribute_name=attribute_name,
                                               claim_name=claim_name,
                                               claim_type=wrapped_json_type,
                                               mapper_type=wrapped_mapper_type,
                                               add_to_id_token=add_to_id_token,
                                               add_to_access_token=add_to_access_token,
                                               add_to_user_info=add_to_user_info,
                                               multi_valued=multi_valued,
                                               aggregate_attribute_values=aggregate_attribute_values
                                               )

            return self.agent_stub.addProtocolMapper(request, metadata=metadata)

        except Exception:
            logger.exception("Error occurred in add_protocol_mapper, probably due to invalid parameters")
            raise
